gnc/sub8_perception/src/sub8_pcl/torpedo_board.cc

```diff
@@ -0,0 +1,146 @@
+  cv::medianBlur(hsv_channels[1], sat_blurred, 5);
+  // cv::imshow("hue median filtered", hue_blurred);
+
+  // Histogram parameters
+  int hist_size = 256;    // bin size
+  float range[] = { 0, 255 };
+  const float *ranges[] = { range };
+
+  // Segment out torpedo board
+  cv::Mat threshed_hue, threshed_sat, segmented_board;
+  int target_yellow = 20;
+  int target_saturation = 180;
+  sub::statistical_image_segmentation(hue_blurred, threshed_hue, hist_size, ranges,
+                                      target_yellow, "Hue", 6, 0.5, 0.5);
+  sub::statistical_image_segmentation(sat_blurred, threshed_sat, hist_size, ranges,
+                                      target_saturation, "Saturation", 6, 0.1, 0.1);
+  segmented_board = threshed_hue / 2.0 + threshed_sat / 2.0;
+  cv::threshold(segmented_board, segmented_board, 255*0.75, 255, cv::THRESH_BINARY);
+  
+#ifdef VISUALIZE
+  // cv::imshow("segmented board", segmented_board);
+#endif
+
+  // cv::resize(segmented_board, segmented_board, cv::Size(0,0), 2, 2);
+  dest = segmented_board;
+}
+
+
+void Sub8TorpedoBoardDetector::find_board_corners(const cv::Mat &segmented_board, sub::Contour &corners){
+  cv::Mat convex_hull_working_img = segmented_board.clone();
+  std::vector<sub::Contour> contours, connected_contours;
+  sub::Contour convex_hull, corner_points;
+
+  /// Find contours
+  cv::findContours( convex_hull_working_img, contours, CV_RETR_EXTERNAL, 
+                    CV_CHAIN_APPROX_SIMPLE);
+
+  // Put longest 2 contours at beginning of "contours" vector
+  std::partial_sort(contours.begin(), contours.begin() + 2, contours.end(), sub::larger_contour);
+
+  // Connect yellow pannels
+  cv::Point pt1 = sub::contour_centroid(contours[0]);
+  cv::Point pt2 = sub::contour_centroid(contours[1]);
+  convex_hull_working_img = cv::Mat::zeros(convex_hull_working_img.size(), CV_8UC1);
+  cv::drawContours(convex_hull_working_img, contours, 0, cv::Scalar(255));
+  cv::drawContours(convex_hull_working_img, contours, 1, cv::Scalar(255));
+  cv::line(convex_hull_working_img, pt1, pt2, cv::Scalar(255));
+  cv::findContours( convex_hull_working_img, connected_contours, CV_RETR_EXTERNAL, 
+                    CV_CHAIN_APPROX_SIMPLE);
+
+  // Put longest contour at beginning of "connected_contours" vector
+  std::partial_sort(contours.begin(), contours.begin() + 1, contours.end(), sub::larger_contour);
+
+  // Find convex hull of connected panels
+  cv::convexHull(connected_contours[0], convex_hull);
+  cv::Point board_centroid = sub::contour_centroid(convex_hull);
+  ROS_INFO((boost::format("convex hull size = %1%") % convex_hull.size()).str().c_str() );
+  int poly_pts = convex_hull.size();
+  const int max_iterations = 50;
+  int total_iterations = 0;
+  double epsilon = 1;
+  double epsilon_step = 0.5;
+  bool corners_success = false;
+  for(int i = 0; i < max_iterations; i++){
+    cv::approxPolyDP(convex_hull, convex_hull, epsilon, true);
```

Comment:
Yea, 1 is conservative for epsilon but if it's not enough to reduce the number of points in a curve then it gets larger as needed. It's a good compromise between staying close to the original curve and not taking too many iterations to complete or getting stuck.